learning/tour-of-beam/frontend/lib/pages/tour/widgets/playground.dart

```diff
@@ -59,7 +65,45 @@ class PlaygroundWidget extends StatelessWidget {
           right: 30,
           child: Row(
             children: [
-              RunOrCancelButton(playgroundController: playgroundController),
+              RunOrCancelButton(
+                playgroundController: playgroundController,
+                beforeCancel: (runner) {
+                  PlaygroundComponents.analyticsService.sendUnawaited(
+                    CancelRunAnalyticsEvent(
+                      snippetContext:
+                          TobEventSnippetContext.fromEventSnippetContext(
+                        eventSnippetContext: runner.eventSnippetContext!,
+                        unitId: tourNotifier.currentUnitId,
+                      ),
+                      duration: runner.elapsed!,
+                    ),
+                  );
+                },
+                beforeRun: () {
+                  PlaygroundComponents.analyticsService.sendUnawaited(
+                    RunAnalyticsEvent(
+                      snippetContext:
```

Comment:
1. Create `EventAbstractContext` that only has `toJson()` abstract method.
2. Make `EventSnippetContext` extend it.
3. Rename event fields `snippetContext -> context` or `eventContext`?
4. Make `TobEventContext get eventContext` in `TourNotifier`.